lib/netext/resolver.go

```diff
@@ -0,0 +1,140 @@
+/*
+ *
+ * k6 - a next-generation load testing tool
+ * Copyright (C) 2020 Load Impact
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package netext
+
+import (
+	"math/rand"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/loadimpact/k6/lib"
+)
+
+// MultiResolver returns all IP addresses for the given host.
+type MultiResolver func(host string) ([]net.IP, error)
+
+// Resolver is an interface that returns DNS information about a given host.
+type Resolver interface {
+	LookupIP(host string) (net.IP, error)
+}
+
+type resolver struct {
+	resolve    MultiResolver
+	strategy   lib.DNSStrategy
+	rrm        *sync.Mutex
+	rand       *rand.Rand
+	roundRobin map[string]uint8
+}
+
+type cacheRecord struct {
+	ips        []net.IP
+	lastLookup time.Time
+}
+
+type cacheResolver struct {
+	resolver
+	ttl   time.Duration
+	cm    *sync.Mutex
+	cache map[string]cacheRecord
+}
+
+// NewResolver returns a new DNS resolver. If ttl is not 0, responses
+// will be cached per host for the specified period. The IP returned from
+// LookupIP() will be selected based on the given strategy.
+func NewResolver(actRes MultiResolver, ttl time.Duration, strategy lib.DNSStrategy) Resolver {
+	r := rand.New(rand.NewSource(time.Now().UnixNano())) // nolint: gosec
```

Comment:
This wouldn't be safe to use from multiple goroutines concurrently, see https://golang.org/pkg/math/rand/#NewSource
> Unlike the default Source used by top-level functions, this source is not safe for concurrent use by multiple goroutines.

Reply:
So I guess it wasn't panicking because the resolver is used globally?

Anyway, I reused the same mutex for `r.roundRobin` in 3149b4d9. Let me know if this is sufficient.